src/estimagic/parameters/nonlinear_constraints.py

```diff
@@ -0,0 +1,468 @@
+        _transformation_type = "subtract_lb"
+    else:
+        # the external constraint can only be transformed by duplication (stacking)
+        _transformation_type = "stack"
+    return _transformation_type
+
+
+# ======================================================================================
+# Checks
+# ======================================================================================
+
+
+def _check_validity_and_return_evaluation(c, params, skip_checks):
+    """Check that nonlinear constraints are valid.
+
+    Returns:
+        constaint_eval: Evaluation of constraint at params, if skip_checks if False,
+            else None.
+
+    """
+    # ==================================================================================
+    # check functions
+    # ==================================================================================
+
+    if "func" not in c:
+        raise ValueError(
+            "Constraint needs to have entry 'fun', representing the constraint "
+            "function."
+        )
+    if not callable(c["func"]):
+        raise ValueError("Entry 'fun' in nonlinear constraints has be callable.")
+
+    if "derivative" in c and not callable(c["derivative"]):
+        raise ValueError("Entry 'jac' in nonlinear constraints has be callable.")
+
+    # ==================================================================================
+    # check bounds
+    # ==================================================================================
+
+    is_equality_constraint = "value" in c
+
+    if is_equality_constraint:
+        if "lower_bounds" in c or "upper_bounds" in c:
+            raise ValueError(
+                "Only one of 'value' or ('lower_bounds', 'upper_bounds') can be "
+                "passed to a nonlinear constraint."
+            )
+
+    if not is_equality_constraint:
+        if "lower_bounds" not in c and "upper_bounds" not in c:
+            raise ValueError(
+                "For inequality constraint at least one of ('lower_bounds', "
+                "'upper_bounds') has to be passed to the nonlinear constraint."
+            )
+
+    if "lower_bounds" in c and "upper_bounds" in c:
+        if not np.all(np.array(c["lower_bounds"]) <= np.array(c["upper_bounds"])):
+            raise ValueError(
+                "If lower bounds need to less than or equal to upper bounds."
+            )
+
+    # ==================================================================================
+    # check selector
+    # ==================================================================================
+
+    if "selector" in c:
+        if not callable(c["selector"]):
+            raise ValueError(
+                f"'selector' entry needs to be callable in constraint {c}."
+            )
+        else:
+            try:
+                c["selector"](params)
+            except Exception:
+                raise InvalidFunctionError(
```

Comment:
Here I am indifferent between `InvalidFunctionError` and `InvalidConstraintError`.